### PR TITLE
Add schema migration for new tables

### PR DESCRIPTION
**Pull request by jschaul**
* schema migration for remote users in local conversations; and users -> remote conversations
* partially implements adding a remote user to a local conversation: cassandra queries are made use of (data is actually inserted), however checks are missing and that endpoint is still hidden from actual usage by using a `/i/` prefix.
* augments `OtherMember` to also give back a qualified user Id; this means that querying for a conversation (`get /conversations/:cnv`) supports showing remote users. For this, golden tests had to be adjusted, which accounts for some of the noise of this PR.
* removing a remote member is implemented at the cql query level but never actually used nor tested. That is a futurework.
* There may be some minor conflicts with the work in #1482 (we needed *some* instances to make servant work); use the instances of #1482 when resolving conflicts.
* Adds scaffolding for internal endpoints in galley with servant.
* Adds lots of FUTUREWORKs.

See the comment in https://github.com/wireapp/wire-server/pull/1485/files#diff-259be1b8909f30421ab02e60765d6e90ae68426b95c3aa8303c57945dc84dd37R467-R481 for some next steps that can be implemented once this PR is merged.

### CODE REVIEW

File: libs/types-common/src/Data/Qualified.hs [hunk 1]
```diff
@@ -97,6 +99,8 @@ data Qualified a = Qualified
   }
   deriving stock (Eq, Ord, Show, Generic, Functor)
 
+type RemoteUserId = Tagged "remote" (Qualified UserId)
```
Comment (akshaymankar) [inline libs/types-common/src/Data/Qualified.hs:102]:
```suggestion
type Remote a = Tagged "remote" (Qualified a)
```

File: services/galley/schema/src/V49_ReAddRemoteIdentifiers.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+-- This migration replaces the remote identifiers deleted in migration 48 with separate tables.
+-- This change occurs because we decided to stop using opaque Ids. Instead, we'll be explict with remote identifiers.
+-- Since two backends may have a conversation or user with the same UUID
+-- (whether by chance or maliciously so), this change guarantees we don't
+-- accidentally override information about a conversation on one backend by
+-- information about a conversation on another backend.
+migration :: Migration
+migration = Migration 49 "Add tables for remote users in local conversations or remote conversations in local membership" $ do
+  -- The user_remote (similar to the user) table answers the question:
```
Comment (pcapriotti) [inline services/galley/schema/src/V49_ReAddRemoteIdentifiers.hs:35]: should be `user_remote_conv`

File: services/galley/src/Galley/Data.hs [hunk 3]
```diff
@@ -749,25 +780,32 @@ memberLists convs = do
...
-addMembersWithRole :: MonadClient m => UTCTime -> ConvId -> (UserId, RoleName) -> ConvMemberAddSizeChecked (List1 (UserId, RoleName)) -> m (Event, List1 LocalMember)
-addMembersWithRole t c orig mems = addMembersUncheckedWithRole t c orig (fromMemberSize mems)
+addMembersWithRole :: MonadClient m => UTCTime -> ConvId -> (UserId, RoleName) -> ConvMemberAddSizeChecked -> m (Event, [LocalMember], [RemoteMember])
+addMembersWithRole t c orig mems = addMembersUncheckedWithRole t c orig (sizeCheckedLocals mems) (sizeCheckedRemotes mems)
 
 -- | Add members to a local conversation, all as admins.
 -- Please make sure the conversation doesn't exceed the maximum size!
-addMembersUnchecked :: MonadClient m => UTCTime -> ConvId -> UserId -> List1 UserId -> m (Event, List1 LocalMember)
-addMembersUnchecked t conv orig usrs = addMembersUncheckedWithRole t conv (orig, roleNameWireAdmin) ((,roleNameWireAdmin) <$> usrs)
+addMembersUnchecked :: MonadClient m => UTCTime -> ConvId -> UserId -> List1 UserId -> m (Event, [LocalMember])
```
Comment (pcapriotti) [inline services/galley/src/Galley/Data.hs:796]: Shouldn't this be called `addLocalMembersUnchecked`?

File: services/galley/src/Galley/Data/Queries.hs [hunk 4]
```diff
@@ -283,6 +283,34 @@ updateMemberHidden = "update member set hidden = ?, hidden_ref = ? where conv =
...
+
+-- FUTUREWORK: actually make use of these cql statements.
+insertRemoteUserConv :: PrepQuery W (UserId, Domain, ConvId) ()
+insertRemoteUserConv = "insert into user_remote_conv (user, conv_remote_domain, conv_remote_id) values (?, ?, ?)"
+
+deleteRemoteUserConv :: PrepQuery W (UserId, Domain, ConvId) ()
+deleteRemoteUserConv = "delete from user_remote_conv where user = ? and conv_remote_domain = ? and conv_remote_id = ?"
+
+selectRemoteUserConvs :: PrepQuery R (Identity UserId) (Domain, ConvId)
+selectRemoteUserConvs = "select conv_remote_domain, conv_remote_id from user_remote_conv where user = ? order by conv_remote_domain"
```
Comment (pcapriotti) [inline services/galley/src/Galley/Data/Queries.hs:312]: The naming here is confusing, and it doesn't match the table name. Maybe these functions should be called `xxxUserRemoteConv`?

File: services/galley/test/integration/API.hs [hunk 5]
```diff
@@ -857,6 +863,31 @@ leaveConnectConversation = do
   let c = fromMaybe (error "invalid connect conversation") (cnvId <$> responseJsonUnsafe bdy)
   deleteMember alice alice c !!! const 403 === statusCode
 
+-- This test adds a non existent remote user and expects a 200. This is of
```
Comment (pcapriotti) [inline services/galley/test/integration/API.hs:866]: I would rephrase as "adds a non-existent remote user *to a conversation*". I was confused at first, thinking this was creating a remote user.

File: services/galley/src/Galley/API/Public.hs [hunk 6]
```diff
@@ -152,6 +152,10 @@ instance
 instance ToSchema a => ToSchema (Headers ls a) where
   declareNamedSchema _ = declareNamedSchema (Proxy @a)
 
+-- FUTUREWORK: Make a PR to the servant-swagger package with this instance
+instance ToSchema Servant.NoContent where
+  declareNamedSchema _ = declareNamedSchema (Proxy @())
+
```
Comment (pcapriotti) [inline services/galley/src/Galley/API/Public.hs:158]: This is not normally necessary, because servant-swagger has a more specific instance for `Verb ... NoContent`. Here we need it because `NoContent` is inside `UVerb`, but I'm not sure this is the best way.

I think this is fine for now, but for the purposes of submitting a PR, I think the proper way to support `NoContent` in `UVerb` is to have some more specific instances for `ToSwagger` that do not necessarily require `ToSchema`. I have done something similar for `HasServer` and `HasClient` in my [servant PR](https://github.com/haskell-servant/servant/pull/1420).

File: libs/wire-api/src/Wire/API/Event/Conversation.hs [hunk 7]
```diff
@@ -92,6 +93,10 @@ data Event = Event
   }
   deriving stock (Eq, Show, Generic)
 
+-- TODO: Merge work from https://github.com/wireapp/wire-server/pull/1482
```
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/Event/Conversation.hs:96]:
```suggestion
-- TODO: Merge work from https://github.com/wireapp/wire-server/pull/1506
```

**Review by pcapriotti — APPROVED** [review]
I have taken the liberty to merge develop, and then remove JSON instances that were covered by schemas.

Looks good to me. I've added a few minor inline comments.

**Review by smatting — APPROVED** [review]

File: libs/types-common/src/Data/Qualified.hs [hunk 8]
```diff
@@ -96,6 +101,16 @@ data Qualified a = Qualified
   }
   deriving stock (Eq, Ord, Show, Generic, Functor)
 
+-- | A type to differentiate between generally Qualified values, and values
+-- where it is known if they are coming from a Remote backend or not.
+-- Use 'toRemote' or 'partitionRemoteOrLocalIds\'' to get Remote values and use
+-- 'unTagged' to convert from a Remote value back to a plain Qualified one.
+type Remote a = Tagged "remote" (Qualified a)
```
[the remark this answers is not in the document]
Comment (jschaul) [reply libs/types-common/src/Data/Qualified.hs:108]: That is a very good question. It was @akshaymankar 's idea, so maybe he can give an answer?

File: libs/types-common/src/Data/Qualified.hs [hunk 9]
```diff
@@ -108,6 +123,9 @@ partitionRemoteOrLocalIds localDomain = foldMap $ \qualifiedId ->
     then (mempty, [qUnqualified qualifiedId])
     else ([qualifiedId], mempty)
 
+partitionRemoteOrLocalIds' :: Foldable f => Domain -> f (Qualified a) -> ([Remote a], [a])
+partitionRemoteOrLocalIds' localDomain xs = first (fmap toRemote) $ partitionRemoteOrLocalIds localDomain xs
+
```
[the remark this answers is not in the document]
Comment (jschaul) [reply libs/types-common/src/Data/Qualified.hs:128]: The unprimed function can be removed; however it's in use in brig and we didn't want to remove/change a range more things in this PR. I will add a FUTUREWORK.

File: libs/types-common/src/Data/Qualified.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/types-common/src/Data/Qualified.hs:108]: There is no advantage, do you think we should stick with newtypes?

File: libs/wire-api/src/Wire/API/Conversation/Member.hs [hunk 10]
```diff
@@ -169,26 +170,31 @@ newtype MutedStatus = MutedStatus {fromMutedStatus :: Int32}
...
+  { omQualifiedId :: Qualified UserId,
     omService :: Maybe ServiceRef,
     omConvRoleName :: RoleName
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform OtherMember)
+  deriving (FromJSON, ToJSON, S.ToSchema) via Schema OtherMember
+
+omId :: OtherMember -> UserId
+omId = qUnqualified . omQualifiedId
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/Conversation/Member.hs:182]: This function could be a cause of logical errors. If we really want a utility like this, we should call it `omLocalId` or something. If we just want to save on integration tests tripping over lack of this function, perhaps we could write this function in integration tests?

File: services/galley/src/Galley/Data.hs (hunk 3 above)
Comment (jschaul) [reply services/galley/src/Galley/Data.hs:796]: Good idea.

File: services/galley/src/Galley/Data/Queries.hs (hunk 4 above)
Comment (jschaul) [reply services/galley/src/Galley/Data/Queries.hs:312]: yes, you're right. Fixed.

File: libs/wire-api/src/Wire/API/Conversation/Member.hs (hunk 10 above)
Comment (jschaul) [reply libs/wire-api/src/Wire/API/Conversation/Member.hs:181]: I removed the function entirely, as it's only used in the schema instance below (I replaced all other occurrences)
